doc/decisions/1_in_discussion/copy_on_write.md

```diff
@@ -0,0 +1,474 @@
+
+This is already implemented for the MMAP cache, so the implementation should be straightforward:
+Do the same COW duplications as done for MMAP but with a different flag.
+
+@kodebach wrote:
+
+> What I wanted to say is that mmap already does COW, so we can reuse the code and probably the flag.
+> If there is some code that is only needed for mmap and not COW, we could make mmap set two flags one for mmap and one for the general COW code.
+
+For the metadata, however, also COW KeySets might be needed (at least with the current API).
+Example:
+
+```c
+keyCopy (cow, key, ELEKTRA_CP_COW);
+KeySet * cowMeta = keyMeta (cow);
+ksAppendKey (cowMeta, keyNew ("meta:/whatever", KEY_VALUE, "abc", KEY_END));
+ksRemoveByName (cowMeta, "meta:/type");
+```
+
+**Pros:**
+
+- Elektra doesn't require MMAP
+
+**Cons:**
+
+- Lifetime of a copied COW key MUST be less than the key it was copied from.
+  We can not track how many keys point to the same data and name this way, so we can only free data and name if the key does not have the COW flag.
+  If the original key gets deleted, using a COW key that points to the same data and name will lead to corrupt data.
+  The same is true for updating values of the original key.
+
+  This is only problematic if we want to use COW for keys outside of `KDB`.
+  If it is only for use within `KDB`, especially for usage as internal cache and in change tracking, we could always guarantee that the original keys are going to last as long as the `KDB` instance.
+  However, we need to document for the users of Elektra that keys returned from `kdbGet` are only valid until `kdbClose`.
+  If they want to continue using them afterwards, they'd have to deep copy them.
+
+  Triggering the delete problem:
+
+  ```c
+  Key * originalKey;
+  Key * copiedKey;
+  keyCopy (copiedKey, originalKey, ELEKTRA_CP_COW);
+
+  assert (keyString (copiedKey) == keyString (originalKey));
+  assert (keyName (copiedKey) == keyName (originalKey));
+
+  keyDel (originalKey);
+
+  keyString (copiedKey); // Error! Original value has been deleted. Pointer to data in copiedKey points to freed memory
+  keyName (copiedKey);   // Error! Original name has been deleted.
+  ```
+
+  Triggering the update problem:
+
+  ```c
+  Key * originalKey;
+  Key * copiedKey;
+  keyCopy (copiedKey, originalKey, ELEKTRA_CP_COW);
+
+  assert (keyString(copiedKey) == keyString(originalKey));
+  keySetString (originalKey, "new value!");
+
+  keyString(copiedKey); // Error! Original value has been deleted. Pointer to data in copiedKey points to potentially freed memory
+  ```
+
+  The same problems in principle exist for `mmapstorage` where `kdbSet` frees (`munmap`) the keyset.
+  We can still let users access the flag `ELEKTRA_CP_COW`, we just need to clearly document what is forbidden.
+  Maybe set the `KEY_FLAG_RO_VALUE` on the original key, so that the API itself detects the error.
+  There is, however, no flag for `keyDel` that we could set.
+
+#### Changes to `libelektra-core`
```

Comment:
Unless anybody sees significant advantages in this approach over the "full-blown COW" approach, I think we can consider this a "discarded idea" and mark it as such. Unlike the `mmap` idea above, here it is not obvious that the idea has downsides, or why it is inferior to the "full-blown COW". It sounds like this idea would solve everything.